File: static/account-manager/BrowserProfiles.js
(() => {
  "use strict";

  const esc = value => String(value ?? "").replace(/[&<>"']/g, ch => ({
    "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;"
  }[ch]));

  async function render(root, options = {}) {
    root.innerHTML = '<div class="am-empty">Проверяю browser profiles...</div>';

    try {
      const payload = await options.api("/api/account-manager/browser/profiles");
      paint(root, payload, options);
    } catch (error) {
      root.innerHTML = '<div class="am-error">Не удалось загрузить browser profiles: ' + esc(error.message) + '</div>';
    }
  }

  function paint(root, payload, options) {
    const browser = payload.browser || {};
    const profiles = payload.profiles || [];

    root.innerHTML =
      '<div class="am-page-stack">' +
        '<section class="am-hero">' +
          '<div class="am-hero-copy">' +
            '<span class="am-eyebrow">LOCAL BROWSER SESSIONS</span>' +
            '<h2>Browser Profiles</h2>' +
            '<p>Каждый Account открывается в отдельном Chrome/Edge user-data-dir. Логины и подтверждения выполняются вручную в обычном браузере.</p>' +
          '</div>' +
          '<div class="am-hero-actions">' +
            '<span class="am-connection ' + (browser.available ? "ok" : "off") + '">' +
              '<span class="am-connection-dot"></span>' +
              (browser.available ? "browser ready" : "browser not found") +
            '</span>' +
            '<button type="button" class="am-btn" data-refresh>↻ Refresh</button>' +
          '</div>' +
        '</section>' +

        (!browser.available
          ? '<section class="am-notice warn"><div class="am-notice-icon">!</div><div><strong>Chrome/Edge не найден</strong><p>' +
              esc(browser.message || "Установи Chrome") +
              '</p></div></section>'
          : '<section class="am-panel am-browser-meta"><div><span>Detected browser</span><b>' + esc(browser.browser || "—") + '</b></div></section>') +

        '<section class="am-panel am-table-panel">' +
          '<div class="am-browser-list" data-browser-list></div>' +
        '</section>' +
      '</div>';

    const list = root.querySelector("[data-browser-list]");
    if (!profiles.length) {
      list.innerHTML = '<div class="am-empty compact">Сначала создай Account.</div>';
    } else {
      list.innerHTML = profiles.map(profile => row(profile, browser.available)).join("");
    }

    root.querySelector("[data-refresh]")?.addEventListener("click", () => render(root, options));

    root.querySelectorAll("[data-open-browser]").forEach(button => {
      button.addEventListener("click", async () => {
        const accountId = Number(button.dataset.accountId);
        const target = button.dataset.openBrowser || "home";
        button.disabled = true;
        try {
          const result = await options.api("/api/account-manager/accounts/" + accountId + "/browser/open", {
            method: "POST",
            body: JSON.stringify({ target })
          });
          options.toast(result.message || "Browser открыт");
          await render(root, options);
        } catch (error) {
          options.toast("Browser: " + error.message);
          button.disabled = false;
        }
      });
    });

    root.querySelectorAll("[data-close-browser]").forEach(button => {
      button.addEventListener("click", async () => {
        const accountId = Number(button.dataset.accountId);
        button.disabled = true;
        try {
          const result = await options.api("/api/account-manager/accounts/" + accountId + "/browser/close", {
            method: "POST",
            body: "{}"
          });
          options.toast(result.message || "Browser закрыт");
          await render(root, options);
        } catch (error) {
          options.toast("Browser: " + error.message);
          button.disabled = false;
        }
      });
    });
  }

  function row(profile, browserAvailable) {
    const label = "Account " + String(profile.account_number || 0).padStart(2, "0");
    return '<article class="am-browser-row">' +
      '<div class="am-browser-profile">' +
        '<div class="am-avatar">' + esc((profile.display_name || profile.username || label).slice(0, 2).toUpperCase()) + '</div>' +
        '<div><strong>' + esc(label) + '</strong><span>' + esc(profile.display_name || profile.username || profile.email || "Profile not configured") + '</span></div>' +
      '</div>' +
      '<div class="am-browser-path">' +
        '<span>PROFILE PATH</span><b>' + esc(profile.browser_profile_path || "—") + '</b>' +
      '</div>' +
      '<div class="am-browser-state">' +
        '<span class="am-state ' + (profile.running ? "free" : "") + '">' + (profile.running ? "RUNNING" : "CLOSED") + '</span>' +
      '</div>' +
      '<div class="am-browser-actions">' +
        '<button type="button" class="am-btn subtle small" data-open-browser="home" data-account-id="' + Number(profile.account_id) + '" ' + (browserAvailable ? "" : "disabled") + '>Open</button>' +
        '<button type="button" class="am-btn subtle small" data-open-browser="tiktok" data-account-id="' + Number(profile.account_id) + '" ' + (browserAvailable ? "" : "disabled") + '>TikTok</button>' +
        '<button type="button" class="am-btn subtle small" data-open-browser="instagram" data-account-id="' + Number(profile.account_id) + '" ' + (browserAvailable ? "" : "disabled") + '>Instagram</button>' +
        '<button type="button" class="am-btn subtle small" data-open-browser="youtube" data-account-id="' + Number(profile.account_id) + '" ' + (browserAvailable ? "" : "disabled") + '>YouTube</button>' +
        '<button type="button" class="am-btn danger small" data-close-browser data-account-id="' + Number(profile.account_id) + '">Close</button>' +
      '</div>' +
    '</article>';
  }

  window.PostingTTIIBrowserProfiles = { render };
})();

File: app/account_creator/api/browser_profiles.py
from __future__ import annotations

from typing import Literal

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from app.account_creator import database
from app.account_creator.services.browser_service import (
    BrowserProfileError,
    close_profile,
    is_running,
    open_profile,
    status as browser_status,
)
from app.account_creator.services.instagram_service import (
    is_running as is_instagram_running,
    stop_instagram_browser,
)

router = APIRouter(prefix="/api/account-manager", tags=["account-manager-browser"])


class BrowserOpenRequest(BaseModel):
    target: Literal["home", "tiktok", "instagram", "youtube"] = "home"


@router.get("/browser/status")
def status():
    return browser_status()


@router.get("/browser/profiles")
def profiles():
    accounts = database.list_accounts()
    return {
        "browser": browser_status(),
        "profiles": [
            {
                "account_id": account["id"],
                "account_number": account["account_number"],
                "display_name": account["display_name"],
                "username": account["username"],
                "email": account["email"],
                "status": account["status"],
                "browser_profile_path": account["browser_profile_path"],
                "running": is_running(account["id"]) or is_instagram_running(account["id"]),
                "instagram_running": is_instagram_running(account["id"]),
            }
            for account in accounts
        ],
    }


@router.post("/accounts/{account_id}/browser/open")
def open_browser(account_id: int, body: BrowserOpenRequest):
    account = database.get_account(account_id)
    if not account:
        raise HTTPException(404, "Account not found")

    try:
        return open_profile(
            account_id,
            account["browser_profile_path"],
            body.target,
        )
    except BrowserProfileError as exc:
        raise HTTPException(503, str(exc)) from exc


@router.post("/accounts/{account_id}/browser/close")
def close_browser(account_id: int):
    account = database.get_account(account_id)
    if not account:
        raise HTTPException(404, "Account not found")
    stop_instagram_browser(account_id)
    return close_profile(account_id)
